pkg/maps/nat/nat_mock.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nat
+
+import (
+	"github.com/cilium/cilium/pkg/bpf"
+)
+
+// NatMockMap implements the NatMap interface and can be used for unit tests.
+type NatMockMap struct {
+	Entries []NatMapRecord
```

Comment:
This is nice to see as an example of the kind of mocking we would want to be able to properly unit-test the core of the daemon. I wonder if you have thought about how we might centralize these mocks so that we have a single implementation you can use to instantiate a memory-backed "BPF" map in a generic way without having to duplicate the similar base functionality in each package?

FWIW we have a `pkg/testutils` that hosts a bunch of things like this.

Separately, we've also been pondering about whether it makes sense to have such a mock framework as part of the `cilium/ebpf` repository. Without examples like these, we haven't really known what such mocks would need to look like just yet.

Reply:
I did think about centralising them, but I wasn't sure there was enough interest at this time and preferred not to wander too far in that direction, apart from reusing the `dumpAndRead()` bit. But yeah, definitely something we could consider for the future. Most of the code for NAT and CT mock maps is very close, it should not be too hard to figure out something. Depending on what other map types or use cases you have in mind, it could add some challenge, of course.

Reply:
One mock to rule them all! ...well yeah I guess just mainly the ones we're interacting with from Cilium userspace which are primarily hash-or-similar: LRU, LPM, hashmap. I don't think we even at the moment need to care *too* much about the fact that lookups for LPM work differently, as BPF LPM map lookups are only done from the (Envoy) proxy and not from cilium-agent [citation needed].